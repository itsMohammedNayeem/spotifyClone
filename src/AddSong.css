.addsong {
  flex: 0.6;
  flex-wrap: wrap;
  margin: 0 20px;
}

.addsong__inputContainer {
  background-color: white;
  padding: 20px;
  padding-bottom: 20px;
  border-radius: 10px;
  margin-bottom: 20px;
}

.addsong__comment {
  padding: 10px;
  display: flex;
  align-items: center;
  border-radius: 25px;
  height: 15px;
  width: auto;
  color: gray;
  background-color: #eef3f8;
  margin: 10px;
}

.addsong__comment > input {
  outline: none;
  border: none;
  background: none;
  font-weight: 400;
  margin-left: 5px;
  width: 60%;
}

.addsong__buttons {
  display: flex;
  flex-wrap: wrap;
  padding-right: 10px;
  padding-left: 10px;
  padding-top: 20px;
  justify-content: space-between;
}

.form-control {
  margin-top: 1rem;
}

.example label:first-child,
.example .form-control:first-child {
  margin-top: 0;
}

.input-feedback {
  color: red;
  margin-top: 0.25rem;
}

.text-sm {
  font-size: 0.8rem;
}

.space-between {
  justify-content: space-between;
}

.items-center {
  align-items: center;
}

.text-xs {
  font-size: 0.75rem;
}

.error {
  color: red;
}

.success {
  color: green;
}

input:invalid,
select:invalid,
textarea:invalid {
  border: 2px solid #ff7d87;
  box-shadow: none;
}

.invalid input[type="email"],
.invalid input[type="text"],
.invalid input[type="number"] {
  border-color: red;
}

.invalid .feedback {
  color: red;
}

.valid input[type="email"],
.valid input[type="text"],
.valid input[type="number"] {
  border-color: green;
}

.valid .feedback {
  color: green;
}

.flex {
  display: flex;
  flex-wrap: wrap;
}

.example {
  padding: 1rem;
  border: 1px solid #e2e2e4;
  border-radius: 4px;
}

.example label:first-child,
.example .form-control:first-child {
  margin-top: 0;
}

@media (prefers-color-scheme: dark) {
  .example {
    border: 1px solid #444444;
  }
}

.form-control {
  margin-top: 1rem;
}
