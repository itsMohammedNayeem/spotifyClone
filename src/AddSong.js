import React, { useState } from "react";
import SaveIcon from "@material-ui/icons/Save";
import MusicNoteIcon from "@material-ui/icons/MusicNote";
import HistoryIcon from "@material-ui/icons/History";
import RecordVoiceOverIcon from "@material-ui/icons/RecordVoiceOver";
import AlbumIcon from "@material-ui/icons/Album";
import CloseIcon from "@material-ui/icons/Close";
import { Link, Prompt } from "react-router-dom";
import "./AddSong.css";
import { useFormik, FormikProvider, Form, useField } from "formik";
import * as Yup from "yup";
import { IconButton } from "@material-ui/core";

function AddSong(props) {
  const [title, setTitle] = useState("");
  const [movie, setMovie] = useState("");
  const [duration, setDuration] = useState("");
  const [artist, setArtist] = useState("");

  const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

  const TextInputLiveFeedback = ({ ...props }) => {
    const [field, meta] = useField(props);

    const [didFocus, setDidFocus] = useState(false);
    const handleFocus = () => setDidFocus(true);
    const showFeedback =
      (!!didFocus && field.value.trim().length > 2) || meta.touched;

    return (
      <div
        className={`form-control ${
          showFeedback ? (meta.error ? "invalid" : "valid") : ""
        }`}
      >
        <div className="addsong__comment">
          {props.id === "title" ? (
            <MusicNoteIcon style={{ color: "E99497" }} />
          ) : null}
          {props.id === "movie" ? (
            <AlbumIcon style={{ color: "F3C583" }} />
          ) : null}
          {props.id === "duration" ? (
            <HistoryIcon style={{ color: "E8E46E" }} />
          ) : null}
          {props.id === "artist" ? (
            <RecordVoiceOverIcon style={{ color: "B3E283" }} />
          ) : null}
          <input
            {...props}
            {...field}
            aria-describedby={`${props.id}-feedback ${props.id}-help`}
            onFocus={handleFocus}
          />
          {showFeedback ? (
            <div
              id={`${props.id}-feedback`}
              aria-live="polite"
              className="feedback text-sm"
            >
              {meta.error ? meta.error : "âœ“"}
            </div>
          ) : null}
        </div>
      </div>
    );
  };

  const Example = () => {
    const formik = useFormik({
      initialValues: {
        title: "",
        movie: "",
        duration: "",
        artist: "",
      },
      onSubmit: async ({ title, movie, duration, artist }) => {
        await sleep(500);
        setTitle(title);
        setMovie(movie);
        setDuration(duration);
        setArtist(artist);
        props.addSongHandler({ title, movie, duration, artist });
        setTitle("");
        setMovie("");
        setDuration("");
        setArtist("");
        props.history.push("/list");
      },
      validationSchema: Yup.object({
        title: Yup.string()
          .min(2, "Must be at least 2 characters")
          .max(200, "Must be less  than 200 characters")
          .required("* Required")
          .matches(/^[a-zA-Z0-9]+/, "Cannot contain special characters"),
        movie: Yup.string()
          .min(2, "Must be at least 2 characters")
          .max(500, "Must be less  than 500 characters")
          .required("* Required")
          .matches(/^[a-zA-Z0-9]+/, "Cannot contain special characters"),
        duration: Yup.string()
          .min(3, "Must be at least 3 characters")
          .max(5, "Must be less  than 5 characters")
          .required("* Required")
          .matches(/^[0-9]+:/, "Can only contain number"),
        artist: Yup.string()
          .min(2, "Must be at least 2 characters")
          .max(100, "Must be less  than 100 characters")
          .required("* Required")
          .matches(/^[a-zA-Z0-9]+/, "Cannot contain special characters"),
      }),
    });

    return (
      <FormikProvider value={formik}>
        <Form>
          <TextInputLiveFeedback id="title" name="title" type="text" />
          <TextInputLiveFeedback id="movie" name="movie" type="text" />
          <TextInputLiveFeedback id="duration" name="duration" type="text" />
          <TextInputLiveFeedback id="artist" name="artist" type="text" />
          <div className="addsong__buttons">
            <IconButton type="submit">
              <SaveIcon style={{ color: "89CFF0" }} />
            </IconButton>
            <Link to="/list" style={{ textDecoration: "none" }}>
              <IconButton>
                <CloseIcon style={{ color: "FF7779" }} />
              </IconButton>
            </Link>
          </div>
          <Prompt
            when={
              !!formik.values.title ||
              !!formik.values.movie ||
              !!formik.values.duration ||
              !!formik.values.artist
            }
            message="Do you want to leave without saving data?"
          />
        </Form>
      </FormikProvider>
    );
  };

  return (
    <div className="addsong">
      <div className="addsong__inputContainer">
        <h2>Add Song</h2>
        <Example />
      </div>
    </div>
  );
}

export default AddSong;
