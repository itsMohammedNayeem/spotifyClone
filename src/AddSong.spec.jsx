import React from "react";
import { mount, shallow } from "enzyme";
import AddSong from "./AddSong";

describe("AddSong", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<div />);
  });

  test("render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("displays div", () => {
    expect(wrapper.find("div").length).toEqual(1);
  });
});
