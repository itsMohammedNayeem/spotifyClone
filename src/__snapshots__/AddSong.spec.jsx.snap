// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddSong render correctly 1`] = `ShallowWrapper {}`;
